### ISSUE TITLE
Feature Request: Add Hard-Time option for Log Chain Event

### ISSUE THREAD

**Issue by alexolivan** [ElvishArtisan/rivendell#815]
Hi... I've messing with the Reports, and I've realized a problem with my ToH first break:

As I've told that I've to somehow 'over book' to be safe while programming logs. So far, so good, but that causes that the chain log event always executes beyond 00:00, already within the next day wallclock time.
The consequence of this, as I read the reports, is that the 'hard time' setting of the first event in the log seems to be modified in order to be executed asap... which is certainly not a bad idea, but not necessarily what is always wanted.

Investigating on the mailing list, I've come across several people asking about this.
The problem with the answers I found is that, basically, you have to have a dedicated 23->00/day-trailing version of any clock you need at that time, that includes some make-next event, maybe a small silence cart, prior to the chainlog event, to ensure it gets executed before 00:00 ... and I feel this somehow 'too hacky'  

Asking my partners of radio automation dptm on how this is solved in Win systems they use, they told me this should be done with a 'COMMAND' (Kind of RML/MACRO + RDCatch approach equivalent) that simply makes the next day log to be loaded at ToH (no timing alteration/recalculation) and executed... easy and straightforward.
But I think this can't be done with a macro... is it?

I thing it would be easier to just have a 'blue-line/T version' of the Chain Log event on the log.
The idea would be to add the option in RDAdmin, so you could opt to enable Hard Time Chain Log option (optionally, toghether with a time, so the chain log is executed at that time, instead of 23:59:59.9) and no recalculation of first hard-time event happens, but just plain, simple, wallclock 00:00 synced, next-day log loading.

Cheers.

**Comment by dklann** [comment]
Hi @alexolivan , I have worked around this limitation with a Macro that calls a [shell script](https://github.com/opensourceradio/ram/blob/master/usr/local/bin/log-chain-timed-play). The Macro (and shell script) is launched from RDCatch **after** generating the Log. The script uses _rdclilogedit_ to modify the _LOG CHN_ event to PLAY with a Hard Start (make next). (The linked script is a ZSH script, but there is actually nothing inherently ZSH-specific in the script.) The Script assumes there is only one Event containing the text "LOG CHN".

I concur, it would be helpful for this to be an user-selectable option in RDAdmin.

**Comment by alexolivan** [comment]
HI... really, REALLY interesting... thank you very much for the point!

So, from your approach I see that, in reality, no 'hacky' thing is required at all... You just treated LOG CHN event just as a regular one ... and it worked! So, in reality, 99,% of the thing is already there (just miss the GUI part as you mention)

Me too I'm using a script that generates 'next day' log automatically. However, once I've to return to CLI to have things done, I tend to consider GUI not 'being the intended way'... and i drop it completelly. In this case, I'm simply using crontab.

I think it would be very much useful (Having in mind  the average user that is not skilled in CLI) to be also be able to trigger a log generation somehow from GUI, so, the whole log management cycle could be handled from GUI, without CLI skills.
Maybe the LOG CHAIN Event could have an option that would also insert, somewhere previously in the log, an 'special' event for next day log generation.

Cheers.

**Comment by alexolivan** [comment]
Hi.

I'm starting to experiment thanks to your script.
I've managed to have mine changin the event to hard time with play transition... However, I have to closely inspect what's the behaviour of RDAirplay... I've to see it.

I will start by doing like you at 23:59:30, and see what happens (I'm wondering what would happen if I use 00:00:00)

I would like to have an RML command and be able to control the log chaining from RDCatch without scripting.
However, the advantage of having the logchain as a trailing event in the log is a very nice way to ensure that, no matter what, upon reaching log, the next one will be loaded.

Cheers!

**Comment by alexolivan** [comment]
Found RDAirplay dead sunday morning.
Next day log was not loaded.
The Log chain event, hard-timed by my script at 23:59:57 , appeared to be executed but apparently the loading operation took to much or something happened... my guess is that the brief stopping of the main log machine halted or aborted the next day log load.

I was concerned about having to halt the Main Log machine in order to have the next day log back in sync at 00:00:00 ToH (I've been advised to never, ever, stop the main log machine while in automatic/unattended operation) and now I got the confirmation this is bad idea indeed.
Today it has worked again... the next day log started synced at 00:00:00 ... but I'm staring to feel that, unless I'm missing something on the manual, the system usability is seriously compromised if something so esential can't be done cleanly as part of the bussiness-logic, without hacks and tricks.

Cheers.

**Comment by alexolivan** [comment]
> It can. See the ['Load Log' ['LL'] RML](http://opsguide.rivendellaudio.org/html/sect.rml.commands.html#sect.rml.load_log__ll_). I've seen this used quite successfully at a number of sites.

Hi Fred! ... for LL, what I'm missing is how can i use RML language to load 'Next Day Log'.

While the software provides facilities to generate 'next day' log, and to chain a log into 'next day' log, it seems to lack the way to point to 'next day' log using LL, AFAIK, It requires to specify the target log...

More simply, what I miss is to see that the suite can have the next-day log loaded and ready to go, by the 00:00:00 mark, so the first event in the log (T00:00:00 PLAY) goes just in time (like the rest of 23 ToH marks do), without stopping, without advanced/hacky tricks. This should be natural and straightforward. 
Surely, I feel I'm missing something obvious...

Cheers!

**Comment by WLGW-Ops** [comment]
@alexolivan - I realize this is an older "issue" so you may have figured this out already, regardless my station uses an RML macro as Fred mentioned above. This is laughably easy once you see it. 

1. Create a macro cart set to execute asynchronously with the following RML command structure: **_LL 1 WLGW-%m%d!_** (obviously replace WLGW with your service name)
2. Create an RDCatch event to load your macro cart at 00:00:00. You can create an event for each host if you have multiple Rivendell instances (i.e. backup air studio, production studio, etc).
3. Make sure your log clock for the 23:00 hour is over-scheduled by a couple minutes or so (we generally do this for each hour so as to keep our hourly news feed on schedule.

FYI for log generation, we have a simple crontab entry that invokes rdlogmanager 5 min before midnight:
**_55 23 * * * rdlogmanager -g -s WLGW_**

![Screenshot from 2022-12-14 11-53-09](https://user-images.githubusercontent.com/119166967/207672122-7c21664b-4380-4595-be6d-682b18addaaa.png)

![Screenshot from 2022-12-14 11-53-41](https://user-images.githubusercontent.com/119166967/207672142-1ecd212b-031f-4a4b-8b6f-fae3d0ec0f4b.png)

Let me know if this helps!

Best Regards,
Brian
WLGW Operations

**Comment by alexolivan** [comment]
> 1. %m%d

Ouch! that's the piece I was missing!
Thank you for this clue! very very interesting... This opens the way again to do further testing, since your approach, toghether with that clue really seems to put everything into place.

Now but, before testing, I'm wondering whether you combine this RDcatch logical approach with the Log Chain event. I mean...do you get rid of the log chain event for the log settings? or you combine both (I think that the log chain event, because of overschedue, will simpley never reach, but may act as a last resource / safety mean of chaining log)

Thank you very much!

**Comment by WLGW-Ops** [comment]
@alexolivan,

We skip the Log Chain event entirely. Our log generation and loading is literally the process I outlined above. You may be able to use Log Chain in conjunction with the above as a redundancy, but I suspect the chain event may execute regardless of whether your log load would trigger beforehand. If you're going to try using both approaches, I'd do so in a backup instance / non-air production studio before you roll it out in a live environment (unless someone replies with better experience and guidance on this approach). There shouldn't be additional features required given the options above. Good luck and let me know how it works for you!

Best Regards,
Brian
WLGW Operations

**Comment by alexolivan** [comment]
Thank you! 
Will feed back if I find some issue.

My intention is to close this Feature Request entirely as it seems out of the question if I manage to use wildcards to match logs in RML macro language (I bet it is documented and I overlooked it)
In my scenario, my logs do require year wildcard too (%y I guess), since they're like: servicename-%m%d%y ... but I expect this to work.

Hopefully, if someone in the future fails as I did on putting a so-elementary piece of the puzzle in place, may find this thread anyways.
